src/TestFramework/PhpUnit/Coverage/IndexXmlCoverageParser.php

```diff
@@ -0,0 +1,293 @@
+use function Safe\preg_replace;
+use function Safe\realpath;
+use function Safe\sprintf;
+use function str_replace;
+
+/**
+ * @internal
+ */
+class IndexXmlCoverageParser
+{
+    /**
+     * Parses the given PHPUnit XML coverage index report (index.xml) to collect the general
+     * coverage data. Note that this data is likely incomplete an will need to be enriched to
+     * contain all the desired data.
+     *
+     * @throws NoLineExecuted
+     * @throws CoverageDoesNotExistException
+     *
+     * @return CoverageFileData[]
+     */
+    public function parse(string $coverageXmlContent): array
+    {
+        $xPath = self::createXPath($coverageXmlContent);
+
+        self::assertHasCoverage($xPath);
+
+        $nodes = self::safeQuery($xPath, '//file');
+
+        $projectSource = self::getProjectSource($xPath);
+
+        $data = [];
+
+        foreach ($nodes as $node) {
+            $relativeFilePath = $node->getAttribute('href');
+
+            $this->processCoverageFile($relativeFilePath, $projectSource, $data);
+        }
+
+        return $data;
+    }
+
+    private static function createXPath(string $coverageContent): DOMXPath
+    {
+        $dom = new DOMDocument();
+        $dom->loadXML(self::removeNamespace($coverageContent));
+
+        return new DOMXPath($dom);
+    }
+
+    /**
+     * Remove namespace to work with xPath without a headache
+     */
+    private static function removeNamespace(string $xml): string
+    {
+        /** @var string $cleanedXml */
+        $cleanedXml = preg_replace('/xmlns=\".*?\"/', '', $xml);
+
+        Assert::string($cleanedXml);
+
+        return $cleanedXml;
+    }
+
+    /**
+     * @throws NoLineExecuted
+     */
+    private static function assertHasCoverage(DOMXPath $xPath): void
+    {
+        $lineCoverage = self::safeQuery($xPath, '/phpunit/project/directory[1]/totals/lines')->item(0);
+
+        if (
+            !($lineCoverage instanceof DOMElement)
+            || ($coverageCount = $lineCoverage->getAttribute('executed')) === '0'
+            || $coverageCount === ''
+        ) {
+            throw NoLineExecuted::create();
+        }
+    }
+
+    /**
+     * @param array<string, CoverageFileData> $data
+     *
+     * @throws CoverageDoesNotExistException
+     */
+    private function processCoverageFile(
+        string $relativeCoverageFilePath,
+        string $projectSource,
+        array &$data
+    ): void {
+        $xPath = self::createXPath(file_get_contents(
+            realpath($projectSource . '/' . $relativeCoverageFilePath)
+        ));
+
+        $sourceFilePath = self::retrieveSourceFilePath($xPath, $relativeCoverageFilePath, $projectSource);
+
+        $linesNode = self::safeQuery($xPath, '/phpunit/file/totals/lines')[0];
+
+        $percentage = $linesNode->getAttribute('percent');
+
+        if ($percentage === '') {
+            $data[$sourceFilePath] = new CoverageFileData();
+
+            return;
+        }
```

Comment:
we do not have any tests for this case. Do you remember a situation where this can happen @maks-rafalko?

Reply:
See https://github.com/infection/infection/commit/1acf6866c4846cabea4fce91ee7165e1bf6022cb#diff-e64a6836ac47da4eaa945539e07f8ec7L78

Why is it being compared to empty string now?

Reply:
Hm looks like a mistake, but looking at the coverage it appears that the result is the same as you would end up in `$coveredLineNodes->length === 0` and return there. So it effectively void another query. Is that the goal?